core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -950,16 +963,17 @@ class DynamicListenerConfig(server: KafkaBroker) extends BrokerReconfigurable wi
     val listenersAdded = newListeners.filterNot(e => oldListenerMap.contains(e.listenerName))
 
     // Clear SASL login cache to force re-login
-    if (listenersAdded.nonEmpty || listenersRemoved.nonEmpty)
-      LoginManager.closeAll()
-
-    server.socketServer.removeListeners(listenersRemoved)
-    if (listenersAdded.nonEmpty)
-      server.socketServer.addListeners(listenersAdded)
-
-    server match {
-      case kafkaServer: KafkaServer => kafkaServer.kafkaController.updateBrokerInfo(kafkaServer.createBrokerInfo)
-      case _ =>
+    LoginManager.closeAll()
```

Comment:
Do we still want to do this even if we reject the configuration change?

Reply:
So, what is confusing about this code is that the same key can be registered for multiple `Reconfigurable` objects. In this case, `DynamicListenerConfig` (created in `DynamicBrokerConfig#addReconfigurables`) is registered for `max.connections` (and the other properties listed in `DynamicListenerConfig#ReconfigurableConfigs`) but SocketServer is also registered for those properties. If you notice, `DynamicListenerConfig` doesn't do anything with `max.connections`, or even look at it.

So *unless* you are adding or removing a listener, the only effect of `DynamicListenerConfig` is to call `LoginManager#closeAll`. This is needed for some SASL things. For example, if you change `sasl.jaas.config` I suspect you need to do this.

I suppose this raises the question of if we should be doing something to ensure the SocketServer's reconfiguration happens before or after the call to `LoginManager#closeAll`. Given that this is an existing code path, I left it the way it was. Maybe we should file a follow-up JIRA about this and pull in some people who know more about SASL.

I will also add a comment here about this.

Reply:
I fixed this so that we only call this if an endpoint was added or removed, which is how it was previously.